env/fs_posix.cc

```diff
@@ -1041,6 +1041,96 @@ class PosixFileSystem : public FileSystem {
   }
 #endif  // ROCKSDB_IOURING_PRESENT
 
+  virtual IOStatus Poll(std::vector<IOHandle*>& /*io_handles*/,
+                        size_t /*min_completions*/) override {
+    IOStatus io_s;
+#if defined(ROCKSDB_IOURING_PRESENT)
+    // Step 1: io_uring_queue_init.
+    struct io_uring* iu = nullptr;
+    if (thread_local_io_urings_) {
+      iu = static_cast<struct io_uring*>(thread_local_io_urings_->Get());
+      if (iu == nullptr) {
+        iu = CreateIOUring();
+        if (iu != nullptr) {
+          thread_local_io_urings_->Reset(iu);
+        }
+      }
+    }
+
+    // Init failed, platform doesn't support io_uring.
+    if (iu == nullptr) {
+      return IOStatus::NotSupported("Poll");
+    }
+
+    while (true) {
+      // Step 2: io_uring_wait_cqe.
+      struct io_uring_cqe* cqe = nullptr;
+      ssize_t ret = io_uring_wait_cqe(iu, &cqe);
+      if (ret) {
+        io_s = IOStatus::IOError("io_uring_wait_cqe() returns " +
+                                 ROCKSDB_NAMESPACE::ToString(ret));
+        if (cqe != nullptr) {
+          io_uring_cqe_seen(iu, cqe);
+        }
+        // TODO akankshamahajan: How to figure out if requested read showed
+        // error or some other request???
+        continue;
+      }
+
+      // Step 3: Populate the request.
+      WrappedAsyncReadRequest* req_wrap =
+          static_cast<WrappedAsyncReadRequest*>(io_uring_cqe_get_data(cqe));
+      Posix_IOHandle* p_iohandle =
+          reinterpret_cast<Posix_IOHandle*>(req_wrap->io_handle);
+
+      if (p_iohandle->iu != iu) {
+        return IOStatus::IOError("");
+      }
+
+      FSReadRequest& req = req_wrap->req;
+      if (cqe->res < 0) {
+        // Request failed.
+        req.result = Slice(req.scratch, 0);
+        req.status = IOStatus::IOError("Req failed");
+      } else {
+        size_t bytes_read = static_cast<size_t>(cqe->res);
+        if (bytes_read == req_wrap->iov.iov_len) {
+          // Request successfull.
+          req.result = Slice(req.scratch, req.len);
+          req.status = IOStatus::OK();
+        } else if (bytes_read == 0) {
+          // No  bytes read.
+          req.result = Slice(req.scratch, 0);
+          req.status = IOStatus::IOError("No bytes read");
+        } else if (bytes_read < req_wrap->iov.iov_len) {
+          // EOF, or partial results.
+          assert(bytes_read + req_wrap->finished_len < req.len);
+          req.result = Slice(req.scratch, req_wrap->finished_len);
+          req.status = IOStatus::OK();
+        } else {
+          // More bytes than requested.
+          req.result = Slice(req.scratch, 0);
+          req.status =
+              IOStatus::IOError("Req returned more bytes than requested");
+        }
+        io_uring_cqe_seen(iu, cqe);
+        req_wrap->cb(req, req_wrap->cb_arg);
+
+        if (p_iohandle->offset == req.offset &&
+            p_iohandle->fd == req_wrap->io_handle->fd) {
```

Comment:
I think what you need instead is to check whether p_iohandle matches one of the elements in the io_handles vector. Once everything in io_handles is matched, the polling is complete.